devnet-sdk/devstack/devtest/testing.go

```diff
@@ -0,0 +1,166 @@
+package devtest
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-service/testlog"
+)
+
+type T interface {
+	CommonT
+
+	// TempDir creates a temporary directory, and returns the file-path.
+	// This directory is cleaned up at the end of the test, and must not be shared between tests.
+	TempDir() string
+
+	// Cleanup runs the given function at the end of the test-scope,
+	// or at the end of the sub-test (if this is a nested test).
+	// This function will clean-up before the package-level testing scope may be complete.
+	// Do not use the test-scope cleanup with shared resources.
+	Cleanup(fn func())
+
+	// Run runs the given function in as a sub-test.
+	Run(name string, fn func(T))
+
+	// Ctx returns a context that will be canceled at the end of this (sub) test-scope,
+	// and inherits the context of the parent-test-scope.
+	Ctx() context.Context
+
+	// Parallel signals that this test is to be run in parallel with (and only with) other parallel tests.
+	Parallel()
```

Comment:
Ideally we'd remove this to force people to use `ParallelT` and not be able to call `SerialT().Parallel()`.

Reply:
We still need it for parallel sub-tests. We can return a reduced interface with `SerialT()`, but it seems a bit over-restrictive to me.

Reply:
ah yeah true.  That's unfortunate. I actually wonder if our test wrapper should automatically mark sub tests as also being parallel.  We get caught out a lot by having `.Parallel()` in the parent test but not every sub-test which winds up with the subtests running sequentially again.

Not a particularly high priority though - it will come up as we keep designing out the DSL and see how this actually plays out.